# Vibe Coding: Real-World Data Cleaning Challenge

## The Mission

You're a Data Analyst at **TechSalary Insights**. Your manager needs answers to critical business questions, but the data is messy. Your job is to clean it and provide accurate insights.

**The catch:** You must figure out how to clean the data yourself. No step by step hints just you, your AI assistant, and real world messy data.

---

## The Dataset: Ask A Manager Salary Survey 2021

**Location:** `../Week-02-Pandas-Part-2-and-DS-Overview/data/Ask A Manager Salary Survey 2021 (Responses) - Form Responses 1.tsv`

This is **real survey data** from Ask A Manager's 2021 salary survey with over 28,000 responses from working professionals. The data comes from this survey: https://www.askamanager.org/2021/04/how-much-money-do-you-make-4.html

**Why this dataset is perfect for vibe coding:**
- Real human responses (inconsistent formatting)
- Multiple currencies and formats  
- Messy job titles and location data
- Missing and invalid entries
- Requires business judgment calls

---

## Your Business Questions

Answer these **exact questions** with clean data. There's only one correct answer for each:

### Core Questions (Required):
1. **What is the median salary for Software Engineers in the United States?** 
2. **Which US state has the highest average salary for tech workers?**
3. **How much does salary increase on average for each year of experience in tech?**
4. **Which industry (besides tech) has the highest median salary?**

### Bonus Questions (If time permits):
5. **What's the salary gap between men and women in tech roles?**
6. **Do people with Master's degrees earn significantly more than those with Bachelor's degrees?**

**Success Criteria:** Your final answers will be compared against the "official" results. Data cleaning approaches can vary, but final numbers should be within 5% of expected values.


---
# Your Work Starts Here

## Step 0: Create Your Plan
**Before writing any code, use Cursor to create your todo plan. Then paste it here:**

## My Data Cleaning Plan

### Phase 1: Initial Data Loading & Exploration
- [ ] Load the TSV file with proper separator
- [ ] Check the shape of the dataset (rows, columns)
- [ ] Examine column names and data types
- [ ] Look at first/last few rows to understand structure
- [ ] Check for missing values across all columns
- [ ] Get summary statistics for numeric columns

### Phase 2: Column Standardization & Cleanup
- [ ] Rename columns to shorter, code-friendly names (e.g., 'What industry do you work in?' → 'industry')
- [ ] Identify which columns are needed for analysis:
  - Timestamp (for date filtering)
  - Age
  - Industry
  - Job title
  - Salary
  - Currency
  - Country
  - State (for US respondents)
  - Years of experience (overall and field-specific)
  - Education level
  - Gender
- [ ] Drop unnecessary columns (additional context fields, etc.)

### Phase 3: Critical Data Cleaning - Currency & Salary
- [ ] **CURRENCY STANDARDIZATION:**
  - Filter to USD only (since questions focus on US data)
  - Or implement currency conversion if needed for non-tech industries
- [ ] **SALARY CLEANING:**
  - Convert salary column to numeric (handle text, commas, etc.)
  - Remove obviously invalid salaries (0, negative, unrealistic values like >$1M or <$10k)
  - Handle missing salary values (drop rows)
  - Check for outliers and decide on reasonable bounds

### Phase 4: Geographic Data Cleaning
- [ ] **COUNTRY STANDARDIZATION:**
  - Standardize US variations: 'US', 'USA', 'United States', 'usa', 'U.S.', etc. → 'US'
  - Keep track of other countries for industry comparison
- [ ] **STATE CLEANING (US only):**
  - Check for state name variations and inconsistencies
  - Handle missing state data
  - Standardize state names (full names vs abbreviations)

### Phase 5: Industry & Job Title Cleaning
- [ ] **INDUSTRY:**
  - Identify exact value for tech/computing industry
  - Create tech industry filter
  - Check unique industries for non-tech analysis
- [ ] **JOB TITLES (for Software Engineer analysis):**
  - Find variations: 'Software Engineer', 'software engineer', 'Software Developer', etc.
  - Create flexible matching (case-insensitive, contains 'software' and 'engineer')
  - Consider related titles: 'Developer', 'SWE', 'Software Dev', etc.
- [ ] **DATA QUALITY - SUSPICIOUS ENTRIES:**
  - Identify suspicious job title + salary combinations
  - Flag or remove obvious joke/fake entries like:
    - Job titles: "bum", "unemployed", "student", "none", etc.
    - Job titles with profanity or gibberish
  - Consider reasonable salary ranges by job category (optional, but helpful)

### Phase 6: Experience & Education Cleaning
- [ ] **YEARS OF EXPERIENCE:**
  - Convert text ranges to numeric values ('5-7 years' → 6, '8-10 years' → 9)
  - Handle '1 year or less' → 0.5 or 1
  - Handle '21-30 years' or '31+ years'
  - Create numeric field_experience column
- [ ] **EDUCATION:**
  - Standardize education levels
  - Create Bachelor's vs Master's groups for comparison

### Phase 7: Gender Data Cleaning (for bonus question)
- [ ] Standardize gender values
- [ ] Handle variations: 'Man', 'Woman', etc.
- [ ] Decide how to handle non-binary/other for gap analysis

### Phase 8: Date Filtering
- [ ] Convert timestamp to datetime
- [ ] Verify all data is from 2021 (survey year)
- [ ] Filter if needed for specific time ranges

### Phase 9: Answer Business Questions
- [ ] **Q1: Median salary for Software Engineers in US**
  - Filter: country='US', job_title contains 'software engineer', currency='USD'
  - Calculate median
- [ ] **Q2: Highest paying US state for tech workers**
  - Filter: country='US', industry='Computing or Tech', currency='USD'
  - Group by state, calculate mean salary
  - Sort and find top state
- [ ] **Q3: Salary increase per year of experience in tech**
  - Filter: tech industry, US, USD
  - Perform regression or correlation between experience and salary
  - Calculate average increase per year
- [ ] **Q4: Highest paying non-tech industry**
  - Filter: exclude tech, USD only
  - Group by industry, calculate median
  - Find highest
- [ ] **Q5 (Bonus): Gender pay gap in tech**
  - Filter: tech roles, US, USD, gender='Man' or 'Woman'
  - Compare median/mean salaries
  - Calculate percentage difference
- [ ] **Q6 (Bonus): Master's vs Bachelor's salary difference**
  - Filter: education level, tech or all industries
  - Compare median salaries
  - Statistical significance test if needed

### Phase 10: Validation & Quality Checks
- [ ] Verify sample sizes for each analysis (ensure sufficient data)
- [ ] Check for and document any assumptions made
- [ ] Sanity check all final numbers (do they make sense?)
- [ ] Document edge cases and how they were handled

### Phase 11: Documentation
- [ ] Document key decisions made during cleaning
- [ ] Note any data quality issues discovered
- [ ] Summarize findings in final section
- [ ] List challenges and solutions

## Step 1: Data Loading and Exploration

Start here! Load the dataset and get familiar with what you're working with.


In [9]:
# Import necessary libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

# ============================================
# PHASE 1: Initial Data Loading & Exploration
# ============================================

# Load the TSV file with proper separator
file_path = '../../Week-02-Pandas-Part-2-and-DS-Overview/data/Ask A Manager Salary Survey 2021 (Responses) - Form Responses 1.tsv'
df = pd.read_csv(file_path, sep='\t')

print("=" * 80)
print("PHASE 1: DATA LOADING & EXPLORATION")
print("=" * 80)

# Check the shape of the dataset
print(f"\n📊 Dataset Shape: {df.shape[0]:,} rows × {df.shape[1]} columns")

# Examine column names and data types
print("\n📋 Column Names and Data Types:")
print("-" * 80)
print(df.dtypes)

# Look at first few rows
print("\n👀 First 5 Rows:")
print("-" * 80)
print(df.head())

# Look at last few rows to understand structure
print("\n👀 Last 5 Rows:")
print("-" * 80)
print(df.tail())

# Check for missing values across all columns
print("\n❓ Missing Values Count:")
print("-" * 80)
missing_counts = df.isnull().sum()
missing_pct = (df.isnull().sum() / len(df)) * 100
missing_df = pd.DataFrame({
    'Missing Count': missing_counts,
    'Percentage': missing_pct
})
print(missing_df[missing_df['Missing Count'] > 0].sort_values('Missing Count', ascending=False))

# Get summary statistics for numeric columns
print("\n📈 Summary Statistics (Numeric Columns):")
print("-" * 80)
print(df.describe())

print("\n✅ Phase 1 Complete!")

PHASE 1: DATA LOADING & EXPLORATION

📊 Dataset Shape: 28,062 rows × 18 columns

📋 Column Names and Data Types:
--------------------------------------------------------------------------------
Timestamp                                                                                                                                                                                                                                object
How old are you?                                                                                                                                                                                                                         object
What industry do you work in?                                                                                                                                                                                                            object
Job title                                                                               

## Step 2: Data Cleaning


In [10]:
# ============================================
# PHASE 2: Column Standardization & Cleanup
# ============================================

print("=" * 80)
print("PHASE 2: COLUMN STANDARDIZATION & CLEANUP")
print("=" * 80)

# First, let's see the original column names
print("\n📝 Original Column Names:")
print("-" * 80)
for i, col in enumerate(df.columns, 1):
    print(f"{i:2}. {col}")

# Rename columns to shorter, code-friendly names
column_mapping = {
    'Timestamp': 'timestamp',
    'How old are you?': 'age',
    'What industry do you work in?': 'industry',
    'Job title': 'job_title',
    'If your job title needs additional context, please clarify here:': 'job_title_context',
    'What is your annual salary? (You\'ll indicate the currency in a later question. If you are part-time or hourly, please enter an annualized equivalent -- what you would earn if you worked the job 40 hours a week, 52 weeks a year.)': 'salary',
    'How much additional monetary compensation do you get, if any (for example, bonuses or overtime in an average year)? Please only include monetary compensation here, not the value of benefits.': 'additional_compensation',
    'Please indicate the currency': 'currency',
    'If "Other," please indicate the currency here: ': 'other_currency',
    'If your income needs additional context, please provide it here:': 'income_context',
    'What country do you work in?': 'country',
    'If you\'re in the U.S., what state do you work in?': 'state',
    'What city do you work in?': 'city',
    'How many years of professional work experience do you have overall?': 'years_experience_overall',
    'How many years of professional work experience do you have in your field?': 'years_experience_field',
    'What is your highest level of education completed?': 'education',
    'What is your gender?': 'gender',
    'What is your race? (Choose all that apply.)': 'race'
}

df = df.rename(columns=column_mapping)

print("\n✅ Columns renamed successfully!")
print("\n📝 New Column Names:")
print("-" * 80)
for i, col in enumerate(df.columns, 1):
    print(f"{i:2}. {col}")

# Identify columns needed for analysis
needed_columns = [
    'timestamp',
    'age',
    'industry',
    'job_title',
    'salary',
    'currency',
    'country',
    'state',
    'years_experience_overall',
    'years_experience_field',
    'education',
    'gender'
]

print(f"\n📊 Columns Needed for Analysis: {len(needed_columns)}")
print("-" * 80)
for col in needed_columns:
    print(f"  ✓ {col}")

# Identify columns to drop
columns_to_drop = [col for col in df.columns if col not in needed_columns]
print(f"\n🗑️  Columns to Drop: {len(columns_to_drop)}")
print("-" * 80)
for col in columns_to_drop:
    print(f"  ✗ {col}")

# Drop unnecessary columns
df_clean = df[needed_columns].copy()

print(f"\n📉 Shape Before Cleanup: {df.shape}")
print(f"📈 Shape After Cleanup: {df_clean.shape}")

print("\n✅ Phase 2 Complete!")
print(f"✨ Cleaned dataset: {df_clean.shape[0]:,} rows × {df_clean.shape[1]} columns")

PHASE 2: COLUMN STANDARDIZATION & CLEANUP

📝 Original Column Names:
--------------------------------------------------------------------------------
 1. Timestamp
 2. How old are you?
 3. What industry do you work in?
 4. Job title
 5. If your job title needs additional context, please clarify here:
 6. What is your annual salary? (You'll indicate the currency in a later question. If you are part-time or hourly, please enter an annualized equivalent -- what you would earn if you worked the job 40 hours a week, 52 weeks a year.)
 7. How much additional monetary compensation do you get, if any (for example, bonuses or overtime in an average year)? Please only include monetary compensation here, not the value of benefits.
 8. Please indicate the currency
 9. If "Other," please indicate the currency here: 
10. If your income needs additional context, please provide it here:
11. What country do you work in?
12. If you're in the U.S., what state do you work in?
13. What city do you work in?


In [11]:
# ============================================
# PHASE 3: Critical Data Cleaning - Currency & Salary
# ============================================

print("=" * 80)
print("PHASE 3: CURRENCY & SALARY CLEANING")
print("=" * 80)

# ----------------
# CURRENCY STANDARDIZATION
# ----------------
print("\n💱 CURRENCY STANDARDIZATION")
print("-" * 80)

# Check currency distribution
print("\n📊 Currency Distribution (Before Filtering):")
print(df_clean['currency'].value_counts())

# Count rows before filtering
rows_before_currency = len(df_clean)

# Filter to USD only (since our questions focus on US data)
df_clean = df_clean[df_clean['currency'] == 'USD'].copy()

rows_after_currency = len(df_clean)
rows_removed = rows_before_currency - rows_after_currency

print(f"\n✅ Filtered to USD only")
print(f"   Rows before: {rows_before_currency:,}")
print(f"   Rows after: {rows_after_currency:,}")
print(f"   Rows removed: {rows_removed:,} ({(rows_removed/rows_before_currency)*100:.1f}%)")

# ----------------
# SALARY CLEANING
# ----------------
print("\n💰 SALARY CLEANING")
print("-" * 80)

# Check current salary data type and sample values
print(f"\n📋 Salary column type: {df_clean['salary'].dtype}")
print(f"\n👀 Sample salary values (first 10):")
print(df_clean['salary'].head(10).tolist())

# Convert salary to numeric (handle any text, commas, etc.)
df_clean['salary'] = pd.to_numeric(df_clean['salary'], errors='coerce')

# Check for missing values after conversion
missing_salaries = df_clean['salary'].isnull().sum()
print(f"\n❓ Missing/Invalid salary values: {missing_salaries:,}")

# Remove rows with missing salaries
rows_before_missing = len(df_clean)
df_clean = df_clean.dropna(subset=['salary']).copy()
rows_after_missing = len(df_clean)
print(f"   Rows removed: {rows_before_missing - rows_after_missing:,}")

# Check salary distribution before filtering outliers
print(f"\n📊 Salary Statistics (Before Outlier Removal):")
print(df_clean['salary'].describe())

# Identify problematic salaries
invalid_zero_negative = df_clean[df_clean['salary'] <= 0]
unrealistically_low = df_clean[(df_clean['salary'] > 0) & (df_clean['salary'] < 15000)]
unrealistically_high = df_clean[df_clean['salary'] > 1000000]

print(f"\n🔍 Salary Quality Check:")
print(f"   Zero or negative: {len(invalid_zero_negative):,}")
print(f"   Unrealistically low (<$15k): {len(unrealistically_low):,}")
print(f"   Unrealistically high (>$1M): {len(unrealistically_high):,}")

if len(unrealistically_high) > 0:
    print(f"\n   Sample high salaries with job titles:")
    high_sample = df_clean[df_clean['salary'] > 1000000][['job_title', 'salary', 'industry']].head(10)
    print(high_sample.to_string())

# Remove invalid salaries
# Using $15,000 as lower bound (minimum wage equivalent)
# Using $1,000,000 as upper bound (very high but not impossible)
rows_before_bounds = len(df_clean)
df_clean = df_clean[(df_clean['salary'] >= 15000) & (df_clean['salary'] <= 1000000)].copy()
rows_after_bounds = len(df_clean)
rows_removed_bounds = rows_before_bounds - rows_after_bounds

print(f"\n✅ Removed invalid salaries (outside $15k-$1M range)")
print(f"   Rows removed: {rows_removed_bounds:,}")

# Final salary statistics
print(f"\n📈 Final Salary Statistics:")
print(df_clean['salary'].describe())

print(f"\n📊 Salary Percentiles:")
percentiles = [10, 25, 50, 75, 90, 95, 99]
for p in percentiles:
    value = df_clean['salary'].quantile(p/100)
    print(f"   {p}th percentile: ${value:,.0f}")

# Check for remaining outliers using IQR method (for information only)
Q1 = df_clean['salary'].quantile(0.25)
Q3 = df_clean['salary'].quantile(0.75)
IQR = Q3 - Q1
lower_fence = Q1 - 1.5 * IQR
upper_fence = Q3 + 1.5 * IQR
outliers_iqr = df_clean[(df_clean['salary'] < lower_fence) | (df_clean['salary'] > upper_fence)]

print(f"\n📍 Statistical Outliers (IQR method - for reference only):")
print(f"   Lower fence: ${lower_fence:,.0f}")
print(f"   Upper fence: ${upper_fence:,.0f}")
print(f"   Outliers detected: {len(outliers_iqr):,} ({(len(outliers_iqr)/len(df_clean))*100:.1f}%)")
print(f"   (Note: Not removing these - they may be legitimate high earners)")

print("\n✅ Phase 3 Complete!")
print(f"✨ Cleaned dataset: {df_clean.shape[0]:,} rows × {df_clean.shape[1]} columns")

PHASE 3: CURRENCY & SALARY CLEANING

💱 CURRENCY STANDARDIZATION
--------------------------------------------------------------------------------

📊 Currency Distribution (Before Filtering):
currency
USD        23374
CAD         1673
GBP         1591
EUR          643
AUD/NZD      504
Other        160
CHF           37
SEK           37
JPY           23
ZAR           16
HKD            4
Name: count, dtype: int64

✅ Filtered to USD only
   Rows before: 28,062
   Rows after: 23,374
   Rows removed: 4,688 (16.7%)

💰 SALARY CLEANING
--------------------------------------------------------------------------------

📋 Salary column type: object

👀 Sample salary values (first 10):
['55,000', '34,000', '62,000', '60,000', '62,000', '33,000', '50,000', '112,000', '45,000', '47,500']

❓ Missing/Invalid salary values: 16,819
   Rows removed: 16,819

📊 Salary Statistics (Before Outlier Removal):
count    6.555000e+03
mean     9.410425e+04
std      1.635568e+05
min      0.000000e+00
25%      5.600000e+0

In [12]:
# ============================================
# PHASE 4: Geographic Data Cleaning
# ============================================

print("=" * 80)
print("PHASE 4: GEOGRAPHIC DATA CLEANING")
print("=" * 80)

# ----------------
# COUNTRY STANDARDIZATION
# ----------------
print("\n🌍 COUNTRY STANDARDIZATION")
print("-" * 80)

# Check current country values
print("\n📊 Top 20 Country Values (Before Standardization):")
print(df_clean['country'].value_counts().head(20))

# Count unique country values
print(f"\n📍 Total unique country values: {df_clean['country'].nunique()}")

# Standardize US variations to 'US'
# Strategy: Use both explicit list AND regex patterns for comprehensive coverage

# First, strip whitespace from all country values
df_clean['country'] = df_clean['country'].str.strip()

# Method 1: Explicit list for common variations
us_variations = [
    'United States',
    'USA',
    'US',
    'usa',
    'U.S.',
    'U.S',
    'us',
    'United States of America',
    'U.S.A.',
    'U.S.A',
    'Usa',
    'united states',
    'United states',
    'UNITED STATES',
    'America',
    'United State',
    'United  States',
    'U. S.',
    'U.s.',
    'u.s.',
    'Uniited States',
    'Unites States'
]

df_clean['country'] = df_clean['country'].replace(us_variations, 'US')

# Method 2: Regex patterns for additional variations
# This catches typos, extra spaces, and other variations
import re

# Pattern explanation:
# (?i) = case insensitive
# ^...$ = must match entire string
# \s* = optional whitespace
# Different patterns for different US variations

regex_patterns = [
    # Matches: "u.s.", "U.S.", "u.s.a.", "U.S.A.", etc. (with or without periods/spaces)
    r'(?i)^u[\s\.]*s[\s\.]*a?[\s\.]*$',
    
    # Matches: "united states", "united state", "unites states", "uniited states", etc.
    r'(?i)^uni[t]*ed\s+states?(\s+of\s+america)?$',
    
    # Matches: "usa", "us" (with optional trailing characters)
    r'(?i)^usa?[\s\-]*$',
    
    # Matches: "america" (careful with this one - only exact match)
    r'(?i)^america$',
    
    # Matches common typos with extra spaces or characters
    r'(?i)^u[\s\.\-]*s[\s\.\-]*$'
]

# Apply regex replacements
for pattern in regex_patterns:
    df_clean['country'] = df_clean['country'].str.replace(pattern, 'US', regex=True)

# Final cleanup: strip again in case regex added spaces
df_clean['country'] = df_clean['country'].str.strip()

print("\n✅ Standardized US country variations")

# Check results
print("\n📊 Top 20 Country Values (After Standardization):")
print(df_clean['country'].value_counts().head(20))

# Count how many US records we have
us_count = (df_clean['country'] == 'US').sum()
total_count = len(df_clean)
us_percentage = (us_count / total_count) * 100

print(f"\n🇺🇸 US Records: {us_count:,} ({us_percentage:.1f}% of dataset)")
print(f"🌎 Non-US Records: {total_count - us_count:,} ({100 - us_percentage:.1f}% of dataset)")

# ----------------
# STATE CLEANING (US only)
# ----------------
print("\n\n🗺️  STATE CLEANING (US ONLY)")
print("-" * 80)

# Filter to US records only for state analysis
df_us = df_clean[df_clean['country'] == 'US'].copy()

print(f"\n📊 Working with {len(df_us):,} US records")

# Check state data quality
print(f"\n📋 State column info:")
print(f"   Total US records: {len(df_us):,}")
print(f"   Records with state data: {df_us['state'].notna().sum():,}")
print(f"   Records missing state: {df_us['state'].isna().sum():,}")
print(f"   Missing state percentage: {(df_us['state'].isna().sum() / len(df_us)) * 100:.1f}%")

# Check unique state values
print(f"\n📍 Unique state values: {df_us['state'].nunique()}")

# Show top states
print("\n📊 Top 20 States (Before Cleaning):")
print(df_us['state'].value_counts().head(20))

# Check for problematic state entries
print("\n🔍 Checking for problematic state entries...")

# Find states that are too long (likely not abbreviations or proper names)
long_states = df_us[df_us['state'].str.len() > 20]['state'].value_counts()
if len(long_states) > 0:
    print(f"\n   States with >20 characters ({len(long_states)} unique):")
    print(long_states.head(10))

# Find states with multiple words (might be "State, State" or other issues)
multi_word_states = df_us[df_us['state'].str.contains(',', na=False)]['state'].value_counts()
if len(multi_word_states) > 0:
    print(f"\n   States with commas ({len(multi_word_states)} unique):")
    print(multi_word_states.head(10))

# Clean state data
# Strip whitespace
df_clean.loc[df_clean['country'] == 'US', 'state'] = df_clean.loc[df_clean['country'] == 'US', 'state'].str.strip()

# Note: We're keeping state names as-is (both full names and abbreviations are acceptable)
# This preserves data while removing obvious issues

print("\n✅ State data cleaned (whitespace removed)")

# Final state statistics
df_us_clean = df_clean[df_clean['country'] == 'US'].copy()
print("\n📊 Top 20 States (After Cleaning):")
print(df_us_clean['state'].value_counts().head(20))

# Decision on missing states
missing_state_count = df_us_clean['state'].isna().sum()
print(f"\n⚠️  Records with missing state data: {missing_state_count:,}")
print("   Decision: Keeping records with missing states for now.")
print("   Note: These will be excluded from state-specific analyses.")

print("\n✅ Phase 4 Complete!")
print(f"✨ Cleaned dataset: {df_clean.shape[0]:,} rows × {df_clean.shape[1]} columns")
print(f"   US records: {(df_clean['country'] == 'US').sum():,}")
print(f"   US records with state: {df_clean[(df_clean['country'] == 'US') & (df_clean['state'].notna())].shape[0]:,}")

PHASE 4: GEOGRAPHIC DATA CLEANING

🌍 COUNTRY STANDARDIZATION
--------------------------------------------------------------------------------

📊 Top 20 Country Values (Before Standardization):
country
United States                2392
USA                          2073
US                            680
United States                 236
Usa                           180
USA                           176
U.S.                          139
United States of America      128
United states                 100
usa                            47
Us                             45
united states                  36
U.S.A.                         16
United States of America       14
U.S                            13
us                             13
United states                   8
U.S.                            8
Usa                             7
United State                    5
Name: count, dtype: int64

📍 Total unique country values: 107

✅ Standardized US country variations

📊 Top 20 Country V

In [13]:
# ============================================
# PHASE 5: Industry & Job Title Cleaning
# ============================================

print("=" * 80)
print("PHASE 5: INDUSTRY & JOB TITLE CLEANING")
print("=" * 80)

# ----------------
# INDUSTRY ANALYSIS
# ----------------
print("\n🏢 INDUSTRY ANALYSIS")
print("-" * 80)

# Check unique industries
print(f"\n📊 Total unique industries: {df_clean['industry'].nunique()}")

# Show top industries
print("\n📊 Top 20 Industries:")
print(df_clean['industry'].value_counts().head(20))

# Identify the tech/computing industry value
print("\n💻 Tech/Computing Industry Identification:")
tech_industry_value = 'Computing or Tech'
tech_count = (df_clean['industry'] == tech_industry_value).sum()
print(f"   Industry name: '{tech_industry_value}'")
print(f"   Tech industry records: {tech_count:,}")

# ----------------
# JOB TITLE ANALYSIS FOR SOFTWARE ENGINEERS
# ----------------
print("\n\n👨‍💻 JOB TITLE ANALYSIS - SOFTWARE ENGINEERS")
print("-" * 80)

# First, let's see what job titles exist in tech industry
tech_jobs = df_clean[df_clean['industry'] == tech_industry_value]['job_title'].value_counts().head(30)
print("\n📋 Top 30 Job Titles in Tech Industry:")
print(tech_jobs)

# Create a flag for Software Engineer roles using flexible matching
# Strategy: Look for variations of "software" AND "engineer/developer"
print("\n🔍 Identifying Software Engineer variations...")

# Convert job titles to lowercase for case-insensitive matching
df_clean['job_title_lower'] = df_clean['job_title'].str.lower().str.strip()

# Define patterns for Software Engineer identification
# Pattern 1: Contains "software" AND ("engineer" OR "developer" OR "dev")
is_swe = (
    (df_clean['job_title_lower'].str.contains('software', na=False)) &
    (
        df_clean['job_title_lower'].str.contains('engineer', na=False) |
        df_clean['job_title_lower'].str.contains('developer', na=False) |
        df_clean['job_title_lower'].str.contains(r'\bdev\b', na=False, regex=True)
    )
)

# Pattern 2: Common abbreviations
is_swe = is_swe | df_clean['job_title_lower'].str.contains(r'\bswe\b', na=False, regex=True)

# Pattern 3: "Software" alone (like "Software Developer", "Software Architect")
is_swe = is_swe | (
    df_clean['job_title_lower'].str.contains('software', na=False) &
    df_clean['job_title_lower'].str.contains('developer|dev |engineer|architect|programmer', na=False, regex=True)
)

df_clean['is_software_engineer'] = is_swe

swe_count = df_clean['is_software_engineer'].sum()
print(f"\n✅ Software Engineer roles identified: {swe_count:,}")

# Show sample of matched titles
print("\n📋 Sample of Matched Software Engineer Titles:")
swe_sample_titles = df_clean[df_clean['is_software_engineer']]['job_title'].value_counts().head(20)
print(swe_sample_titles)

# ----------------
# DATA QUALITY - SUSPICIOUS ENTRIES
# ----------------
print("\n\n🚨 DATA QUALITY - SUSPICIOUS ENTRY DETECTION")
print("-" * 80)

# Define suspicious job title keywords (using word boundaries to avoid false positives)
# Pattern explanation: \b ensures we match complete words only, not substrings
suspicious_patterns = [
    r'\bbum\b',           # "bum" but not "bumblebee"
    r'\bunemployed\b',    # "unemployed" 
    r'^student$',         # ONLY if the entire title is just "student"
    r'\bnone\b',          # "none" but not "nonetheless"
    r'\bn/a\b',           # "n/a"
    r'^na$',              # ONLY if the entire title is just "na"
    r'\btest\b',          # "test" but not "testing" or "contest"
    r'\basdf\b',          # gibberish
    r'\bretired\b',       # "retired"
    r'\bhomemaker\b',     # "homemaker"
    r'\bstay at home\b',  # "stay at home"
    r'\bsahm\b',          # "sahm"
    r'\bxxx\b',           # profanity marker
    r'\bfuck',            # profanity
    r'\bshit',            # profanity
    r'\bcrap\b',          # profanity
    r'\bidiot\b',         # insult
    r'\bstupid\b',        # insult
    r'^nothing$',         # ONLY if entire title is "nothing"
    r'\bno job\b',        # "no job"
    r'\bnojob\b',         # "nojob"
    r'\bseeking\b',       # "seeking" (job seeker)
    r'\blooking for work\b',  # "looking for work"
    r'\bjob seeker\b',    # "job seeker"
    r'\bbetween jobs\b'   # "between jobs"
]

# Check for suspicious entries (case-insensitive)
print("\n🔍 Checking for suspicious job titles...")

# Combine all patterns with OR (|)
suspicious_regex = '|'.join(suspicious_patterns)
suspicious_mask = df_clean['job_title_lower'].str.contains(suspicious_regex, na=False, regex=True)
suspicious_entries = df_clean[suspicious_mask]

print(f"\n⚠️  Suspicious entries found: {len(suspicious_entries):,}")

if len(suspicious_entries) > 0:
    print("\n📋 Sample of Suspicious Entries:")
    print(suspicious_entries[['job_title', 'salary', 'industry']].head(20))
    
    # Show salary stats for suspicious entries
    print("\n💰 Salary stats for suspicious entries:")
    print(suspicious_entries['salary'].describe())

# Check for very short job titles (likely gibberish or incomplete)
short_titles = df_clean[df_clean['job_title'].str.len() <= 2]
print(f"\n📏 Job titles ≤ 2 characters: {len(short_titles):,}")
if len(short_titles) > 0:
    print(short_titles[['job_title', 'salary', 'industry']].head(10))

# Check for missing job titles
missing_titles = df_clean['job_title'].isna().sum()
print(f"\n❓ Missing job titles: {missing_titles:,}")

# Decision: Remove suspicious entries
rows_before_suspicious = len(df_clean)
df_clean = df_clean[~suspicious_mask].copy()
rows_after_suspicious = len(df_clean)
rows_removed_suspicious = rows_before_suspicious - rows_after_suspicious

print(f"\n✅ Removed suspicious job title entries")
print(f"   Rows removed: {rows_removed_suspicious:,}")

# Remove very short titles (likely invalid)
if len(short_titles) > 0:
    rows_before_short = len(df_clean)
    df_clean = df_clean[df_clean['job_title'].str.len() > 2].copy()
    rows_removed_short = rows_before_short - len(df_clean)
    print(f"\n✅ Removed very short job titles (≤2 chars)")
    print(f"   Rows removed: {rows_removed_short:,}")

# Remove missing job titles
if missing_titles > 0:
    rows_before_missing = len(df_clean)
    df_clean = df_clean[df_clean['job_title'].notna()].copy()
    rows_removed_missing = rows_before_missing - len(df_clean)
    print(f"\n✅ Removed missing job titles")
    print(f"   Rows removed: {rows_removed_missing:,}")

# Update the is_software_engineer flag after cleaning
df_clean['job_title_lower'] = df_clean['job_title'].str.lower().str.strip()
is_swe = (
    (df_clean['job_title_lower'].str.contains('software', na=False)) &
    (
        df_clean['job_title_lower'].str.contains('engineer', na=False) |
        df_clean['job_title_lower'].str.contains('developer', na=False) |
        df_clean['job_title_lower'].str.contains(r'\bdev\b', na=False, regex=True)
    )
)
is_swe = is_swe | df_clean['job_title_lower'].str.contains(r'\bswe\b', na=False, regex=True)
df_clean['is_software_engineer'] = is_swe

# ----------------
# FINAL INDUSTRY SUMMARY
# ----------------
print("\n\n📊 FINAL INDUSTRY SUMMARY")
print("-" * 80)

print("\n🏆 Top 15 Industries (After Cleaning):")
print(df_clean['industry'].value_counts().head(15))

tech_count_final = (df_clean['industry'] == tech_industry_value).sum()
swe_count_final = df_clean['is_software_engineer'].sum()

print(f"\n💻 Tech industry records: {tech_count_final:,}")
print(f"👨‍💻 Software Engineer roles: {swe_count_final:,}")

print("\n✅ Phase 5 Complete!")
print(f"✨ Cleaned dataset: {df_clean.shape[0]:,} rows × {df_clean.shape[1]} columns")

PHASE 5: INDUSTRY & JOB TITLE CLEANING

🏢 INDUSTRY ANALYSIS
--------------------------------------------------------------------------------

📊 Total unique industries: 350

📊 Top 20 Industries:
industry
Computing or Tech                       1023
Health care                              536
Education (Higher Education)             506
Nonprofits                               505
Engineering or Manufacturing             472
Accounting, Banking & Finance            437
Government and Public Administration     317
Marketing, Advertising & PR              267
Law                                      245
Business or Consulting                   211
Education (Primary/Secondary)            176
Media & Digital                          164
Insurance                                129
Recruitment or HR                        129
Retail                                   128
Sales                                     91
Property or Construction                  90
Transport or Logistics         

In [20]:
# ============================================
# PHASE 6: Experience & Education Cleaning
# ============================================

print("=" * 80)
print("PHASE 6: EXPERIENCE & EDUCATION CLEANING")
print("=" * 80)

# ----------------
# YEARS OF EXPERIENCE CLEANING
# ----------------
print("\n📅 YEARS OF EXPERIENCE ANALYSIS")
print("-" * 80)

# Check current experience values
print("\n📊 Unique values in 'years_experience_field':")
print(df_clean['years_experience_field'].value_counts().sort_index())

print("\n📊 Unique values in 'years_experience_overall':")
print(df_clean['years_experience_overall'].value_counts().sort_index())

# Function to convert experience text ranges to numeric midpoint
def convert_experience_to_numeric(exp_text):
    """
    Convert experience text ranges to numeric values.
    Examples:
      '5-7 years' -> 6
      '8 - 10 years' -> 9
      '1 year or less' -> 1
      '21 - 30 years' -> 25.5
      '41 years or more' -> 41
    """
    if pd.isna(exp_text):
        return np.nan
    
    exp_text = str(exp_text).strip().lower()
    
    # Handle '1 year or less'
    if '1 year or less' in exp_text or exp_text == '1':
        return 1.0
    
    # Handle '41 years or more' or similar
    if 'or more' in exp_text or '41 years' in exp_text:
        return 41.0
    
    # Extract numbers from ranges like '5-7 years', '8 - 10 years', '2 - 4 years'
    import re
    numbers = re.findall(r'\d+', exp_text)
    
    if len(numbers) == 2:
        # Range like '5-7' -> take midpoint
        return (int(numbers[0]) + int(numbers[1])) / 2
    elif len(numbers) == 1:
        # Single number
        return float(numbers[0])
    else:
        return np.nan

# Apply conversion to both experience columns
print("\n🔄 Converting experience ranges to numeric values...")

df_clean['years_experience_field_num'] = df_clean['years_experience_field'].apply(convert_experience_to_numeric)
df_clean['years_experience_overall_num'] = df_clean['years_experience_overall'].apply(convert_experience_to_numeric)

# Show conversion results
print("\n✅ Experience converted to numeric!")
print("\n📊 Field Experience Statistics:")
print(df_clean['years_experience_field_num'].describe())

print("\n📊 Overall Experience Statistics:")
print(df_clean['years_experience_overall_num'].describe())

# Check for missing values
field_exp_missing = df_clean['years_experience_field_num'].isna().sum()
overall_exp_missing = df_clean['years_experience_overall_num'].isna().sum()

print(f"\n❓ Missing field experience: {field_exp_missing:,} ({(field_exp_missing/len(df_clean))*100:.1f}%)")
print(f"❓ Missing overall experience: {overall_exp_missing:,} ({(overall_exp_missing/len(df_clean))*100:.1f}%)")

# Show sample of conversions
print("\n📋 Sample Conversions (Field Experience):")
sample_df = df_clean[['years_experience_field', 'years_experience_field_num']].drop_duplicates().head(15)
print(sample_df.to_string(index=False))

# ----------------
# EDUCATION CLEANING
# ----------------
print("\n\n🎓 EDUCATION LEVEL ANALYSIS")
print("-" * 80)

# Check current education values
print("\n📊 Education Levels Distribution:")
print(df_clean['education'].value_counts())

# Check for missing education
education_missing = df_clean['education'].isna().sum()
print(f"\n❓ Missing education: {education_missing:,} ({(education_missing/len(df_clean))*100:.1f}%)")

# Standardize education levels
print("\n🔄 Standardizing education levels...")

# Create standardized education categories
def standardize_education(edu_text):
    """
    Standardize education levels into consistent categories.
    """
    if pd.isna(edu_text):
        return 'Unknown'
    
    edu_text = str(edu_text).strip().lower()
    
    # PhD
    if 'phd' in edu_text or 'doctorate' in edu_text:
        return 'PhD'
    
    # Master's degree
    if 'master' in edu_text:
        return "Master's degree"
    
    # College/Bachelor's degree
    if 'college' in edu_text or 'bachelor' in edu_text:
        return "College degree"
    
    # Professional degree (JD, MD, etc.)
    if 'professional' in edu_text:
        return 'Professional degree'
    
    # Some college
    if 'some college' in edu_text:
        return 'Some college'
    
    # High school
    if 'high school' in edu_text:
        return 'High School'
    
    # Default
    return 'Other'

df_clean['education_standard'] = df_clean['education'].apply(standardize_education)

print("\n✅ Education standardized!")
print("\n📊 Standardized Education Distribution:")
print(df_clean['education_standard'].value_counts())

# Create Bachelor's vs Master's comparison groups
df_clean['has_bachelors'] = df_clean['education_standard'] == 'College degree'
df_clean['has_masters'] = df_clean['education_standard'] == "Master's degree"
df_clean['has_phd'] = df_clean['education_standard'] == 'PhD'

bachelors_count = df_clean['has_bachelors'].sum()
masters_count = df_clean['has_masters'].sum()
phd_count = df_clean['has_phd'].sum()

print(f"\n🎓 Education Breakdown:")
print(f"   College/Bachelor's degree: {bachelors_count:,}")
print(f"   Master's degree: {masters_count:,}")
print(f"   PhD: {phd_count:,}")

# Create education level ordering for analysis
education_order = {
    'High School': 1,
    'Some college': 2,
    'College degree': 3,
    "Master's degree": 4,
    'Professional degree': 5,
    'PhD': 6,
    'Other': 0,
    'Unknown': 0
}

df_clean['education_level'] = df_clean['education_standard'].map(education_order)

print("\n✅ Education level numeric scale created (1-6)")

# Show education vs salary preview (for validation)
print("\n💰 Average Salary by Education Level:")
salary_by_edu = df_clean.groupby('education_standard')['salary'].agg(['count', 'mean', 'median']).sort_values('median', ascending=False)
print(salary_by_edu)

print("\n✅ Phase 6 Complete!")
print(f"✨ Cleaned dataset: {df_clean.shape[0]:,} rows × {df_clean.shape[1]} columns")
print(f"\n📊 New columns added:")
print(f"   - years_experience_field_num (numeric)")
print(f"   - years_experience_overall_num (numeric)")
print(f"   - education_standard (standardized categories)")
print(f"   - education_level (numeric 1-6)")
print(f"   - has_bachelors, has_masters, has_phd (boolean flags)")

PHASE 6: EXPERIENCE & EDUCATION CLEANING

📅 YEARS OF EXPERIENCE ANALYSIS
--------------------------------------------------------------------------------

📊 Unique values in 'years_experience_field':
years_experience_field
1 year or less       430
11 - 20 years       1453
2 - 4 years         1396
21 - 30 years        389
31 - 40 years         67
41 years or more       5
5-7 years           1503
8 - 10 years        1162
Name: count, dtype: int64

📊 Unique values in 'years_experience_overall':
years_experience_overall
1 year or less       139
11 - 20 years       2228
2 - 4 years          757
21 - 30 years        762
31 - 40 years        183
41 years or more      24
5-7 years           1115
8 - 10 years        1197
Name: count, dtype: int64

🔄 Converting experience ranges to numeric values...

✅ Experience converted to numeric!

📊 Field Experience Statistics:
count    6405.000000
mean        9.230055
std         6.962347
min         1.000000
25%         3.000000
50%         6.000000
75%  

In [21]:
# ============================================
# PHASE 7: Gender Data Cleaning
# ============================================

print("=" * 80)
print("PHASE 7: GENDER DATA CLEANING")
print("=" * 80)

# ----------------
# GENDER ANALYSIS
# ----------------
print("\n⚧️ GENDER DATA ANALYSIS")
print("-" * 80)

# Check current gender values
print("\n📊 Current Gender Distribution:")
print(df_clean['gender'].value_counts())

# Check for missing gender
gender_missing = df_clean['gender'].isna().sum()
print(f"\n❓ Missing gender: {gender_missing:,} ({(gender_missing/len(df_clean))*100:.1f}%)")

# Show unique gender values
print(f"\n📍 Unique gender values: {df_clean['gender'].nunique()}")
print("\n📋 All unique gender values:")
for val in sorted(df_clean['gender'].dropna().unique()):
    count = (df_clean['gender'] == val).sum()
    print(f"   '{val}': {count:,}")

# ----------------
# STANDARDIZE GENDER VALUES
# ----------------
print("\n\n🔄 STANDARDIZING GENDER VALUES")
print("-" * 80)

# Create standardized gender categories
def standardize_gender(gender_text):
    """
    Standardize gender values into consistent categories.
    """
    if pd.isna(gender_text):
        return 'Unknown'
    
    gender_text = str(gender_text).strip().lower()
    
    # Man/Male
    if gender_text in ['man', 'male', 'm']:
        return 'Man'
    
    # Woman/Female
    if gender_text in ['woman', 'female', 'f']:
        return 'Woman'
    
    # Non-binary
    if 'non-binary' in gender_text or 'nonbinary' in gender_text or 'non binary' in gender_text:
        return 'Non-binary'
    
    # Prefer not to answer / Other
    if 'prefer not' in gender_text or 'other' in gender_text or 'another option' in gender_text:
        return 'Other/Prefer not to answer'
    
    # Default - catch any other variations
    return 'Other/Prefer not to answer'

df_clean['gender_standard'] = df_clean['gender'].apply(standardize_gender)

print("\n✅ Gender values standardized!")
print("\n📊 Standardized Gender Distribution:")
print(df_clean['gender_standard'].value_counts())

# Create flags for binary gender analysis
df_clean['is_man'] = df_clean['gender_standard'] == 'Man'
df_clean['is_woman'] = df_clean['gender_standard'] == 'Woman'
df_clean['is_nonbinary'] = df_clean['gender_standard'] == 'Non-binary'

man_count = df_clean['is_man'].sum()
woman_count = df_clean['is_woman'].sum()
nonbinary_count = df_clean['is_nonbinary'].sum()
other_count = (df_clean['gender_standard'] == 'Other/Prefer not to answer').sum()

print(f"\n⚧️ Gender Breakdown:")
print(f"   Men: {man_count:,} ({(man_count/len(df_clean))*100:.1f}%)")
print(f"   Women: {woman_count:,} ({(woman_count/len(df_clean))*100:.1f}%)")
print(f"   Non-binary: {nonbinary_count:,} ({(nonbinary_count/len(df_clean))*100:.1f}%)")
print(f"   Other/Prefer not to answer: {other_count:,} ({(other_count/len(df_clean))*100:.1f}%)")

# ----------------
# GENDER PAY GAP ANALYSIS DECISION
# ----------------
print("\n\n📊 GENDER PAY GAP ANALYSIS APPROACH")
print("-" * 80)

print("\n💡 Decision for Pay Gap Analysis:")
print("   For binary gender pay gap comparison (Man vs Woman):")
print("   - We will use only 'Man' and 'Woman' categories")
print("   - Non-binary and Other categories will be excluded from binary comparison")
print("   - This ensures clear, statistically valid comparison")
print("   - Non-binary workers can be analyzed separately if needed")

# Show salary preview by gender
print("\n💰 Average Salary by Gender (Preview):")
salary_by_gender = df_clean.groupby('gender_standard')['salary'].agg(['count', 'mean', 'median']).sort_values('median', ascending=False)
print(salary_by_gender)

# Calculate sample sizes for pay gap analysis
binary_gender_sample = df_clean[df_clean['gender_standard'].isin(['Man', 'Woman'])]
print(f"\n📊 Sample size for binary pay gap analysis:")
print(f"   Total records with Man/Woman gender: {len(binary_gender_sample):,}")
print(f"   Men: {(binary_gender_sample['gender_standard'] == 'Man').sum():,}")
print(f"   Women: {(binary_gender_sample['gender_standard'] == 'Woman').sum():,}")
print(f"   Sample size is {'sufficient' if len(binary_gender_sample) > 1000 else 'limited'} for statistical analysis")

# Show tech-specific gender breakdown
tech_gender = df_clean[df_clean['industry'] == 'Computing or Tech'].groupby('gender_standard')['salary'].agg(['count', 'mean', 'median'])
print("\n💻 Tech Industry Gender Breakdown:")
print(tech_gender)

print("\n✅ Phase 7 Complete!")
print(f"✨ Cleaned dataset: {df_clean.shape[0]:,} rows × {df_clean.shape[1]} columns")
print(f"\n📊 New columns added:")
print(f"   - gender_standard (standardized categories)")
print(f"   - is_man, is_woman, is_nonbinary (boolean flags)")

PHASE 7: GENDER DATA CLEANING

⚧️ GENDER DATA ANALYSIS
--------------------------------------------------------------------------------

📊 Current Gender Distribution:
gender
Woman                            4818
Man                              1339
Non-binary                        136
Other or prefer not to answer      66
Name: count, dtype: int64

❓ Missing gender: 46 (0.7%)

📍 Unique gender values: 4

📋 All unique gender values:
   'Man': 1,339
   'Non-binary': 136
   'Other or prefer not to answer': 66
   'Woman': 4,818


🔄 STANDARDIZING GENDER VALUES
--------------------------------------------------------------------------------

✅ Gender values standardized!

📊 Standardized Gender Distribution:
gender_standard
Woman                         4818
Man                           1339
Non-binary                     136
Other/Prefer not to answer      66
Unknown                         46
Name: count, dtype: int64

⚧️ Gender Breakdown:
   Men: 1,339 (20.9%)
   Women: 4,818 (75.2%)
  

In [22]:
# ============================================
# PHASE 8: Date/Timestamp Filtering
# ============================================

print("=" * 80)
print("PHASE 8: DATE/TIMESTAMP FILTERING")
print("=" * 80)

# ----------------
# TIMESTAMP ANALYSIS
# ----------------
print("\n📅 TIMESTAMP ANALYSIS")
print("-" * 80)

# Check current timestamp format
print("\n📋 Sample timestamp values:")
print(df_clean['timestamp'].head(10).tolist())

print(f"\n📊 Timestamp column type: {df_clean['timestamp'].dtype}")

# Convert timestamp to datetime
print("\n🔄 Converting timestamp to datetime...")
df_clean['timestamp_dt'] = pd.to_datetime(df_clean['timestamp'], errors='coerce')

# Check for any conversion failures
timestamp_conversion_failed = df_clean['timestamp_dt'].isna().sum()
print(f"   Conversion failures: {timestamp_conversion_failed:,}")

if timestamp_conversion_failed > 0:
    print("\n⚠️  Sample of failed conversions:")
    failed_timestamps = df_clean[df_clean['timestamp_dt'].isna()]['timestamp'].head(10)
    print(failed_timestamps.tolist())

print("\n✅ Timestamp converted to datetime!")

# ----------------
# DATE RANGE ANALYSIS
# ----------------
print("\n\n📊 DATE RANGE ANALYSIS")
print("-" * 80)

# Extract date components
df_clean['year'] = df_clean['timestamp_dt'].dt.year
df_clean['month'] = df_clean['timestamp_dt'].dt.month
df_clean['day'] = df_clean['timestamp_dt'].dt.day
df_clean['date'] = df_clean['timestamp_dt'].dt.date

# Show date range
min_date = df_clean['timestamp_dt'].min()
max_date = df_clean['timestamp_dt'].max()

print(f"\n📅 Survey Date Range:")
print(f"   Earliest response: {min_date}")
print(f"   Latest response: {max_date}")
print(f"   Duration: {(max_date - min_date).days} days")

# Check year distribution
print("\n📊 Responses by Year:")
year_dist = df_clean['year'].value_counts().sort_index()
print(year_dist)

# Check if all data is from 2021
data_2021 = (df_clean['year'] == 2021).sum()
data_not_2021 = (df_clean['year'] != 2021).sum()

print(f"\n✅ Data from 2021: {data_2021:,} ({(data_2021/len(df_clean))*100:.1f}%)")
if data_not_2021 > 0:
    print(f"⚠️  Data NOT from 2021: {data_not_2021:,} ({(data_not_2021/len(df_clean))*100:.1f}%)")
    print("\n📊 Distribution of non-2021 data:")
    print(df_clean[df_clean['year'] != 2021]['year'].value_counts().sort_index())

# Check month distribution (for 2021 data)
print("\n📊 2021 Responses by Month:")
month_dist_2021 = df_clean[df_clean['year'] == 2021]['month'].value_counts().sort_index()
for month_num, count in month_dist_2021.items():
    month_name = pd.Timestamp(2021, month_num, 1).strftime('%B')
    print(f"   {month_name}: {count:,}")

# ----------------
# DATE FILTERING DECISION
# ----------------
print("\n\n🔍 DATE FILTERING DECISION")
print("-" * 80)

if data_not_2021 > 0:
    print("\n⚠️  Non-2021 data detected!")
    print("\n💡 Decision: Filter to 2021 data only")
    print("   Reason: Survey was conducted in 2021, other years likely errors or late submissions")
    
    rows_before_date_filter = len(df_clean)
    df_clean = df_clean[df_clean['year'] == 2021].copy()
    rows_after_date_filter = len(df_clean)
    rows_removed_date = rows_before_date_filter - rows_after_date_filter
    
    print(f"\n✅ Filtered to 2021 data only")
    print(f"   Rows before: {rows_before_date_filter:,}")
    print(f"   Rows after: {rows_after_date_filter:,}")
    print(f"   Rows removed: {rows_removed_date:,}")
else:
    print("\n✅ All data is from 2021 - no filtering needed!")

# ----------------
# FINAL TIMESTAMP SUMMARY
# ----------------
print("\n\n📊 FINAL TIMESTAMP SUMMARY")
print("-" * 80)

final_min_date = df_clean['timestamp_dt'].min()
final_max_date = df_clean['timestamp_dt'].max()

print(f"\n📅 Final Date Range:")
print(f"   Start: {final_min_date.strftime('%B %d, %Y')}")
print(f"   End: {final_max_date.strftime('%B %d, %Y')}")
print(f"   Total days: {(final_max_date - final_min_date).days}")

# Show response volume by week
print("\n📈 Survey Response Timeline (by week):")
df_clean['week'] = df_clean['timestamp_dt'].dt.isocalendar().week
weekly_responses = df_clean.groupby('week').size().head(10)
print(weekly_responses)

print("\n✅ Phase 8 Complete!")
print(f"✨ Cleaned dataset: {df_clean.shape[0]:,} rows × {df_clean.shape[1]} columns")
print(f"\n📊 New columns added:")
print(f"   - timestamp_dt (datetime)")
print(f"   - year, month, day, date (date components)")
print(f"   - week (week number)")

PHASE 8: DATE/TIMESTAMP FILTERING

📅 TIMESTAMP ANALYSIS
--------------------------------------------------------------------------------

📋 Sample timestamp values:
['4/29/2021 13:59:35', '4/29/2021 13:59:41', '4/29/2021 14:00:09', '4/29/2021 14:00:36', '4/29/2021 14:01:36', '4/29/2021 14:02:38', '4/29/2021 14:03:16', '4/29/2021 14:04:07', '4/29/2021 14:05:03', '4/29/2021 14:05:22']

📊 Timestamp column type: object

🔄 Converting timestamp to datetime...
   Conversion failures: 0

✅ Timestamp converted to datetime!


📊 DATE RANGE ANALYSIS
--------------------------------------------------------------------------------

📅 Survey Date Range:
   Earliest response: 2021-04-29 13:59:35
   Latest response: 2024-07-23 17:51:03
   Duration: 1181 days

📊 Responses by Year:
year
2021    6071
2022     248
2023      53
2024      33
Name: count, dtype: int64

✅ Data from 2021: 6,071 (94.8%)
⚠️  Data NOT from 2021: 334 (5.2%)

📊 Distribution of non-2021 data:
year
2022    248
2023     53
2024     33


## Step 3: Business Questions Analysis

Now answer those important business questions!


In [25]:
# Question 1: What is the median salary for Software Engineers in the United States?

# Filter for Software Engineers in US
q1_data = df_clean[
    (df_clean['country'] == 'US') &
    (df_clean['is_software_engineer'] == True)
]

# Calculate median salary
median_salary = q1_data['salary'].median()

# Answer
print(f"Median salary for Software Engineers in the United States: ${median_salary:,.2f}")
print(f"Sample size: {len(q1_data):,} Software Engineers")

Median salary for Software Engineers in the United States: $141,875.00
Sample size: 234 Software Engineers


In [26]:
# Question 2: Which US state has the highest average salary for tech workers?

# Filter for tech workers in US with valid state data
q2_data = df_clean[
    (df_clean['country'] == 'US') &
    (df_clean['industry'] == 'Computing or Tech') &
    (df_clean['state'].notna())
]

# Calculate average salary by state
state_avg_salary = q2_data.groupby('state')['salary'].mean().sort_values(ascending=False)

# Get the top state
top_state = state_avg_salary.index[0]
top_salary = state_avg_salary.iloc[0]

# Answer
print(f"US state with highest average salary for tech workers: {top_state}")
print(f"Average salary: ${top_salary:,.2f}")
print(f"Sample size: {len(q2_data):,} tech workers across {len(state_avg_salary)} states")

US state with highest average salary for tech workers: California, Oregon
Average salary: $200,000.00
Sample size: 944 tech workers across 50 states


In [27]:
# Question 3: How much does salary increase on average for each year of experience in tech?

# Filter for tech workers with valid experience data
q3_data = df_clean[
    (df_clean['industry'] == 'Computing or Tech') &
    (df_clean['years_experience_field_num'].notna())
]

# Calculate correlation and average increase per year
# Group by years of experience and get average salary
exp_salary = q3_data.groupby('years_experience_field_num')['salary'].mean()

# Calculate the slope (average increase per year) using simple linear relationship
# This is the difference in average salary divided by difference in experience
salary_per_year = (exp_salary.iloc[-1] - exp_salary.iloc[0]) / (exp_salary.index[-1] - exp_salary.index[0])

# Answer
print(f"Average salary increase per year of experience in tech: ${salary_per_year:,.2f}")
print(f"Sample size: {len(q3_data):,} tech workers")

Average salary increase per year of experience in tech: $2,995.19
Sample size: 958 tech workers


In [ ]:

# Question 4: What percentage of respondents work remotely vs. in-office?
#I don't know if this is a trick question but the data doesn't exist for this question


Highest paying industry (besides tech): Corporate Training
Median salary: $280,000.00
Sample size: 5,113 workers across 327 industries


In [29]:
# Question 5: Which industry (besides tech) has the highest median salary?

# Filter for non-tech industries
q5_data = df_clean[df_clean['industry'] != 'Computing or Tech']

# Calculate median salary by industry
industry_median_salary = q5_data.groupby('industry')['salary'].median().sort_values(ascending=False)

# Get the top industry
top_industry = industry_median_salary.index[0]
top_median = industry_median_salary.iloc[0]

# Answer
print(f"Highest paying industry (besides tech): {top_industry}")
print(f"Median salary: ${top_median:,.2f}")
print(f"Sample size: {len(q5_data):,} workers across {len(industry_median_salary)} industries")

Highest paying industry (besides tech): Corporate Training
Median salary: $280,000.00
Sample size: 5,113 workers across 327 industries


In [30]:
# Bonus Question 5: What's the salary gap between men and women in tech roles?

# Filter for tech workers with binary gender (Man or Woman)
q6_data = df_clean[
    (df_clean['industry'] == 'Computing or Tech') &
    (df_clean['gender_standard'].isin(['Man', 'Woman']))
]

# Calculate median salary by gender
gender_salaries = q6_data.groupby('gender_standard')['salary'].median()

men_salary = gender_salaries['Man']
women_salary = gender_salaries['Woman']

# Calculate the gap (as percentage)
pay_gap_percent = ((men_salary - women_salary) / men_salary) * 100

# Answer
print(f"Gender pay gap in tech roles:")
print(f"  Men median salary: ${men_salary:,.2f}")
print(f"  Women median salary: ${women_salary:,.2f}")
print(f"  Pay gap: {pay_gap_percent:.1f}% (women earn {pay_gap_percent:.1f}% less than men)")
print(f"  Sample size: {len(q6_data):,} tech workers (Men: {(q6_data['gender_standard']=='Man').sum():,}, Women: {(q6_data['gender_standard']=='Woman').sum():,})")

print("\n" + "="*80 + "\n")

# Bonus Question 6: Do people with Master's degrees earn significantly more than those with Bachelor's degrees?

# Filter for Bachelor's and Master's degree holders
q7_data = df_clean[
    (df_clean['has_bachelors'] == True) | (df_clean['has_masters'] == True)
]

# Calculate median salary by education level
bachelors_salary = df_clean[df_clean['has_bachelors'] == True]['salary'].median()
masters_salary = df_clean[df_clean['has_masters'] == True]['salary'].median()

# Calculate the difference
salary_difference = masters_salary - bachelors_salary
percent_increase = ((masters_salary - bachelors_salary) / bachelors_salary) * 100

# Answer
print(f"Master's vs Bachelor's degree salary comparison:")
print(f"  Bachelor's degree median salary: ${bachelors_salary:,.2f}")
print(f"  Master's degree median salary: ${masters_salary:,.2f}")
print(f"  Difference: ${salary_difference:,.2f} ({percent_increase:.1f}% higher with Master's)")
print(f"  Sample size: Bachelor's: {df_clean['has_bachelors'].sum():,}, Master's: {df_clean['has_masters'].sum():,}")

Gender pay gap in tech roles:
  Men median salary: $135,000.00
  Women median salary: $112,525.00
  Pay gap: 16.6% (women earn 16.6% less than men)
  Sample size: 921 tech workers (Men: 395, Women: 526)


Master's vs Bachelor's degree salary comparison:
  Bachelor's degree median salary: $74,000.00
  Master's degree median salary: $85,000.00
  Difference: $11,000.00 (14.9% higher with Master's)
  Sample size: Bachelor's: 3,449, Master's: 1,913


## Final Summary

**Summarize your findings here:**

1. **Median salary for Software Engineers in US:** 
2. **Highest paying US state for tech:** California, Oregon
3. **Salary increase per year of experience:** $2,995.19
4. **Remote vs office percentage:** This data doesn't exist
5. **Highest paying non-tech industry:** Corporate Training

**Key insights:**
- Specify where you want the code to be written, sometimes Claude decided I needed to do 2 tasks in one cell and another task across 3
- At least have an idea on how the work should be done, I had to remind Claude to use regex instead of making a list of every possible way to describe "US"

**Challenges faced:**
- The data didn't exist within the survey responses about remote vs in-person work. I argued with Claude until I checked the dataset myself and realized this question was supposed to trip me up
- For the answers to the questions, Claude made the code incredibly bloated. Almost as if we didn't just spend 45 minutes of my time together making the dataset easier to work with. I solved this by instructing the agent to not make things overly complicated and just answer the question with some basic context

**What you learned about vibe coding:**
- You will get very bored waiting for responses sometimes
- I don't think my job is too much at risk anymore
- Don't do it after already trying to watch a toddler at home, you will end up with a broken keyboard
